File: JANL.VB/ExtendedControls/DateTimePickerEx.vb
Imports System.ComponentModel

Public Class DateTimePickerEx
    Inherits DateTimePicker

    Public Sub New()
        MyBase.New()
        MyBase.ShowCheckBox = True
        Format = DateTimePickerFormat.Short
    End Sub

    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
    Public Shadows ReadOnly Property ShowCheckBox As Boolean
        Get
            Return MyBase.ShowCheckBox
        End Get
    End Property

    <Bindable(True), RefreshProperties(RefreshProperties.All)>
    Public Shadows Property Value As Date?
        Get
            Return If(Checked, New Date?(MyBase.Value), Nothing)
        End Get
        Set
            MyBase.Value = If(Value, Date.Now)
            Checked = Value.HasValue
        End Set
    End Property

    Protected Overrides Sub OnLostFocus(e As EventArgs)
        MyBase.OnLostFocus(e)
        If Not Checked Then ForceWrite()
    End Sub

    Protected Overrides Sub OnValidated(e As EventArgs)
        MyBase.OnValidated(e)
    End Sub

    Protected Overrides Sub OnValueChanged(eventargs As EventArgs)
        MyBase.OnValueChanged(eventargs)
    End Sub

    Private Sub ForceWrite()
        For Each B As Binding In DataBindings
            B.WriteValue()
        Next
    End Sub

End Class

File: JANL.WinForms/Controls/DateTimePickerEx.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace JANL
{
    /// <summary>
    /// <see cref="DateTimePicker"/> с рабочей привязкой
    /// </summary>
    public sealed class DateTimePickerEx : DateTimePicker
    {
        /// <summary>
        ///
        /// </summary>
        public DateTimePickerEx()
        {
            base.ShowCheckBox = true;
            Format = DateTimePickerFormat.Short;
        }

        /// <summary>
        ///
        /// </summary>
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public new bool ShowCheckBox => base.ShowCheckBox;

        /// <summary>
        /// Значение
        /// </summary>
        [Bindable(true)]
        [RefreshProperties(RefreshProperties.All)]
        public new DateTime? Value
        {
            get => Checked ? new DateTime?(base.Value) : null;
            set
            {
                base.Value = value ?? DateTime.Now;
                Checked = value.HasValue;
            }
        }

        /// <summary>
        /// Private
        /// </summary>
        protected override void OnLostFocus(EventArgs e)
        {
            base.OnLostFocus(e);
            if (!Checked) { ForceWrite(); }
        }

        /// <summary>
        /// Private
        /// </summary>
        protected override void OnValidated(EventArgs e) => base.OnValidated(e);

        /// <summary>
        /// Private
        /// </summary>
        protected override void OnValueChanged(EventArgs eventargs) => base.OnValueChanged(eventargs);

        private void ForceWrite()
        {
            foreach (Binding B in DataBindings)
            {
                B.WriteValue();
            }
        }
    }
}